docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+selector.
+
+#### Multiple volume types
+
+Many administrators want the ability to dynamically provision multiple volume
+types (meaning, volumes that use different types of plugin) in the same
+cluster.  Some examples:
+
+1.  An administrator hosts their cluster in AWS and wants to provision EBS
+    volumes and GlusterFS volumes using a Gluster Virtual Storage Appliance
+    running in AWS
+2.  An administrator hosts their cluster in OpenStack and wants to provision
+    Cinder and NFS volumes
+
+In order to do this, the identity of the provisioner must be a property of the
+storage class (as opposed to implicit, as it is in the alpha feature).
+
+#### Parameterizing provisioners
+
+It will be common for a cluster to have multiple storage classes which
+leverage the same provisioner to create volumes with different characteristics
+for the different classes.  For example, an administrator running a cluster
+hosted in AWS might want to have two storage classes, `slow` and `fast`, that
+both leverage the same EBS provisioner to provision spinning disk volumes and
+provisioned-iops volumes, respectively.  From this, it follows that:
+
+1.  Storage classes should be able to hold a set of provisioner parameters
+2.  The provisioner should be passed the claim and the details of the storage
+    class it is provisioning for
+
+#### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Analysis
+
+### Summary of alpha functionality
+
+The alpha functionality for dynamic provisioning works as follows:
+
+1.  A cluster has only a single provisioner at a time, dictated by the cloud
+    provider
+2.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+3.  If no volume is found for the claim, and the claim has the annotation
+    `volume.alpha.kubernetes.io/storage-class`, the provisioner is invoked
+    to provision a volume to fill that claim
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
+    that returns a new `Provisioner` instance
+5.  The controller calls the provisioner's `Provision()` method; `Provision`
+    is responsible for provisioning the volume in the cloud provider and
+    returns an `api.PersistentVolume` and an error
+6.  If `Provision` returns an error, the controller stops trying to provision
```

Comment:
The controller will retry provisioning in next periodic sync, currently in 15 seconds.